components/file_system/src/condvar.rs

```diff
@@ -0,0 +1,299 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use futures::{future::FutureExt, pin_mut, select};
+use std::cell::Cell;
+use std::ptr::NonNull;
+use std::sync::Condvar as StdCondvar;
+use std::sync::{Mutex, MutexGuard};
+use std::time::Duration;
+use tokio::sync::Semaphore as TokioSemaphore;
+
+struct DoublyLinkedNode<T> {
+    prev: Cell<Option<NonNull<T>>>,
+    next: Cell<Option<NonNull<T>>>,
+}
+
+impl<T> DoublyLinkedNode<T> {
+    fn new() -> DoublyLinkedNode<T> {
+        DoublyLinkedNode {
+            prev: Cell::new(None),
+            next: Cell::new(None),
+        }
+    }
+}
+
+enum CondvarNode {
+    Sync(StdCondvar, DoublyLinkedNode<CondvarNode>),
+    Async(TokioSemaphore, DoublyLinkedNode<CondvarNode>),
+}
+
+impl CondvarNode {
+    pub fn new_sync() -> CondvarNode {
+        CondvarNode::Sync(StdCondvar::new(), DoublyLinkedNode::new())
+    }
+
+    pub fn new_async() -> CondvarNode {
+        CondvarNode::Async(TokioSemaphore::new(0), DoublyLinkedNode::new())
+    }
+
+    pub fn notify(&self) {
+        match *self {
+            CondvarNode::Sync(ref condv, _) => condv.notify_one(),
+            CondvarNode::Async(ref sem, _) => sem.add_permits(1),
+        }
+    }
+
+    pub fn get_prev(&self) -> Option<NonNull<CondvarNode>> {
+        match *self {
+            CondvarNode::Sync(_, ref node) => node.prev.get(),
+            CondvarNode::Async(_, ref node) => node.prev.get(),
+        }
+    }
+
+    pub fn get_next(&self) -> Option<NonNull<CondvarNode>> {
+        match *self {
+            CondvarNode::Sync(_, ref node) => node.next.get(),
+            CondvarNode::Async(_, ref node) => node.next.get(),
+        }
+    }
+
+    pub fn set_prev(&self, prev: Option<NonNull<CondvarNode>>) {
+        match *self {
+            CondvarNode::Sync(_, ref node) => node.prev.set(prev),
+            CondvarNode::Async(_, ref node) => node.prev.set(prev),
+        }
+    }
+
+    pub fn set_next(&self, next: Option<NonNull<CondvarNode>>) {
+        match *self {
+            CondvarNode::Sync(_, ref node) => node.next.set(next),
+            CondvarNode::Async(_, ref node) => node.next.set(next),
+        }
+    }
+}
+
+/// Un-prioritized conditional variable. Supports both synchronously or
+/// asynchronously waiting on the same instance.
+/// TODO: Prioritized waiting
+/// Maintains multiple linked list for each priority. When notified, instead of
+/// waking up all nodes at once, sort them to a seperate waiting queue and only
+/// wake up the leader.
+#[derive(Debug)]
+pub struct Condvar {
+    head: Cell<Option<NonNull<CondvarNode>>>,
+    tail: Cell<Option<NonNull<CondvarNode>>>,
+}
+
+unsafe impl Send for Condvar {}
+unsafe impl Sync for Condvar {}
```

Comment:
How do you promise the thread-safety? 
By the mutex of `last_refill_time` to make sure `enqueue` and `dequeue` not called in concurrency?

Reply:
Yes, that's why condvar is always paired with a mutex.